src/celeritas/optical/CerenkovGenerator.hh

```diff
@@ -0,0 +1,158 @@
+
+#include "CerenkovDistribution.hh"
+#include "PhotonPrimary.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample Cerenkov photons.
+ */
+class CerenkovGenerator
+{
+  public:
+    // Placeholder struct for tabulated index of refraction values
+    // TODO: remove once we store optical properties
+    struct RefractiveIndex
+    {
+        using Values
+            = Collection<real_type, Ownership::const_reference, MemSpace::native>;
+
+        GenericGridData const& data;
+        Values const& reals;
+    };
+
+  public:
+    // Construct from optical properties and distribution parameters
+    inline CELER_FUNCTION
+    CerenkovGenerator(RefractiveIndex const& refractive_index,
+                      CerenkovDistribution const& dist,
+                      Span<PhotonPrimary> photons);
+
+    // Sample Cerenkov photons from the distribution
+    template<class Generator>
+    inline CELER_FUNCTION void operator()(Generator& rng);
+
+  private:
+    RefractiveIndex const& refractive_index_;
+    CerenkovDistribution const& dist_;
+    Span<PhotonPrimary> photons_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from optical properties and distribution parameters.
+ */
+CELER_FUNCTION
+CerenkovGenerator::CerenkovGenerator(RefractiveIndex const& refractive_index,
+                                     CerenkovDistribution const& dist,
+                                     Span<PhotonPrimary> photons)
+    : refractive_index_(refractive_index), dist_(dist), photons_(photons)
+
+{
+    CELER_EXPECT(refractive_index_.data);
+    CELER_EXPECT(dist_);
+    CELER_EXPECT(photons_.size() == dist_.num_photons);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample Cerenkov photons from the distribution.
+ */
+template<class Generator>
+CELER_FUNCTION void CerenkovGenerator::operator()(Generator& rng)
+{
+    NonuniformGrid<real_type> energy_grid(refractive_index_.data.grid,
+                                          refractive_index_.reals);
+    real_type e_min = energy_grid.front();
+    real_type e_max = energy_grid.back();
+
+    // TODO: rename GenericXsCalculator to something more generic?
+    GenericXsCalculator calc_refractive_index(refractive_index_.data,
+                                              refractive_index_.reals);
+    real_type inv_beta = 2 * constants::c_light
+                         / (dist_.pre.velocity + dist_.post.velocity);
+    real_type cos_max = inv_beta / calc_refractive_index(e_max);
+    real_type sin_max_sq = (1 - cos_max) * (1 + cos_max);
+
+    // Calculate changes over the step
+    Real3 delta_pos = dist_.post.pos;
+    for (int j = 0; j < 3; ++j)
+    {
+        delta_pos[j] -= dist_.pre.pos[j];
+    }
```

Comment:
I think you can include `ArrayOperators` and just do `delta_pos -= dist_.pre.pos`

Reply:
Ah, I was looking for that... thanks!